testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/federation/BackwardsCompatibilityUserStorage.java

```diff
@@ -87,12 +87,12 @@ public UserModel getUserById(String id, RealmModel realm) {
     private UserModel createUser(RealmModel realm, String username) {
         return new AbstractUserAdapterFederatedStorage(session, realm,  model) {
             @Override
-            public String getUsername() {
+            public String doGetUsername() {
```

Comment:
I can see that you'd like to have backwards compatibility. The reasons I did this change is the default methods `getUsername()` and `setUsername(String name)` on `UserModel`:

- Since they are default methods, the `AbstractUserAdapterFederatedStorage` is not abstract anymore
- if you fail to override the `getUsername()` method, you will run into an infinite loop, because the `getUsername()` default method calls the `getFirstAttribute` method, which in AbstractUserAdapterFederatedStorage MUST call the `getUsername` method in order to be able to get the username from somewhere else and keep consistency (i.e. we still want to have `username` accessible as attribute. 

In other words, not introducing this method, everything works completely fine if you override all important methods correctly, but the compiler won't stop you anymore. With the workaround, it's more consistent.

Now, if you say that the AbstractAdapters may not change for backwards compatibility, then that's what needs to happen, so let's see what our options are:

- Delete the method again and swallow the possibility of infinite loops as outlined above (as well as weird naming "Abstract" of non-abstract class)
- Do not have a default implementation of `getUsername()` and `setUsername(...)` on `UserModel`, but require each subclass to provide it. This is a bit inconsistent with the handling of the other methods but that shouldn't really be a problem

The latter is the safest method so I'm going to implement this and let you have a look!

Reply:
Thanks for bringing this. Yes, I think the latter option should be OK. The "getUsername()" and "setUsername()" may not be available by default on the UserModel. At least, I don't see any issues with that right now...

Just FYI. the fact that "Abstract"* class doesn't have any abstract methods, it can still be declared as "abstract" class. But that's really just a detail and maybe not so relevant...

Reply:
Yes I know about abstract - but if you don't have to override the methods that should be overridden, this is a in my view a bug waiting to happen (I've seen this happening).

Anyways, I implemented the approach as indicated.